dimod/discrete/cydiscrete_quadratic_model.pyx

```diff
@@ -109,6 +105,142 @@ cdef class cyDiscreteQuadraticModel:
+
+            if irow[qi] == irow[qi + 1] and icol[qi] >= icol[qi+1]:
+                is_sorted = False
+                break
+
+        if is_sorted:
+            for qi in range(num_interactions):
+                # cython really has a hard time with push_back so we do this
+                # workaround
+                self.bqm_.adj[irow[qi]].first.resize(
+                    self.bqm_.adj[irow[qi]].first.size() + 1)
+                self.bqm_.adj[irow[qi]].first.back().first = icol[qi]
+                self.bqm_.adj[irow[qi]].first.back().second = qdata[qi]
+
+                self.bqm_.adj[icol[qi]].first.resize(
+                    self.bqm_.adj[icol[qi]].first.size() + 1)
+                self.bqm_.adj[icol[qi]].first.back().first = irow[qi]
+                self.bqm_.adj[icol[qi]].first.back().second = qdata[qi]
+
+        else:
+            # this is *much* slower, an alternative would be to make a copy
+            # and then sort but for now let's stick with the simple thing
+            for qi in range(num_interactions):
+                self.bqm_.set_quadratic(irow[qi], icol[qi], qdata[qi])
+
+        # build the adj. This is not really the memory bottleneck so
+        # we can build an intermediate (unordered) set version
+        cdef vector[unordered_set[VarIndex]] adjset
+        adjset.resize(num_variables)
+        u = 0
+        for ci in range(self.bqm_.num_variables()):
+            
+            # we've been careful so don't need ui < case_starts.size() - 1
+            while ci >= self.case_starts_[u+1]:
+                u += 1
+
+            span = self.bqm_.neighborhood(ci)
+
+            v = 0
+            while span.first != span.second:
+                cj = deref(span.first).first
+
+                # see above note
+                while cj >= self.case_starts_[v+1]:
+                    v += 1
+
+                adjset[u].insert(v)
+
+                inc(span.first)
+
+        # now put adjset into adj
+        self.adj_.resize(num_variables)
+        for v in range(num_variables):
+            self.adj_[v].insert(self.adj_[v].begin(),
+                                adjset[v].begin(), adjset[v].end())
+            sort(self.adj_[v].begin(), self.adj_[v].end())
+
+    @classmethod
+    def from_numpy_vectors(cls, starts, ldata, quadratic):
+
+        cdef cyDiscreteQuadraticModel obj = cls()
+
+        try:
+            irow, icol, qdata = quadratic
+        except ValueError:
+            raise ValueError("quadratic should be a 3-tuple")
+
+        # convert to numpy arrays, coercing the types into a simpler set
+        # if necessary
+        index_dtype = np.result_type(starts, irow, icol, np.uint16)
+
+        starts = np.asarray(starts, dtype=index_dtype)
+        ldata = np.asarray(ldata, dtype=obj.dtype)
+        irow = np.asarray(irow, dtype=index_dtype)
+        icol = np.asarray(icol, dtype=index_dtype)
+        qdata = np.asarray(qdata, dtype=obj.dtype)
+
+        if index_dtype == np.uint16:
```

Comment:
No `(u)int8`?

Reply:
A few lines above I coerce it to be at least `np.uint16`

Reply:
Missed that. :+1: